### ISSUE TITLE
Intel Silvermont support

### ISSUE THREAD

**Issue by zumbi** [mozilla/rr#1642]
Trying to run rr on a Silvermont CPU currently throws this error:

[FATAL /usr/src/packages/BUILD/rr-4.0.3/src/PerfCounters.cc:129:get_cpu_microarch() errno: 0 'Success'] 
 -> CPU 0x30670 unknown.

I have looked the manual posted in issue#1607 but I have no clue if the PMU config values are safe to try out as is. Any hint on that regard would be welcome. I am happy to test it here.

**Comment by rocallahan** [comment]
According to http://web.eece.maine.edu/~vweaver/projects/deterministic/deterministic_counters.pdf Atom does not support any deterministic performance counters. You could try using "retired branch instructions" with code 0x5100c4 but I doubt it will work reliably.

**Comment by zumbi** [comment]
Thanks!

Just for the record, hardware is minnowboard, which has an ATOM E3285 CPU.

**Comment by Keno** [comment]
FWIW, I tried to look into this to see if I couldn't get this to work on Scaleway's machines.
I had the most luck with:
```
@@ -61,6 +62,7 @@ struct PmuConfig {

 // XXX please only edit this if you really know what you're doing.
 static const PmuConfig pmu_configs[] = {
+  { IntelSilvermont, "Intel Silvermont", 0x517ec4, 0x5100c0, 0x5301cb, true },
   { IntelSkylake, "Intel Skylake", 0x5101c4, 0x5100c0, 0x5301cb, true },
   { IntelBroadwell, "Intel Broadwell", 0x5101c4, 0x5100c0, 0x5301cb, true },
   { IntelHaswell, "Intel Haswell", 0x5101c4, 0x5100c0, 0x5301cb, true },
@@ -130,6 +132,8 @@ static CpuMicroarch get_cpu_microarch() {
     case 0x406F0:
     case 0x50660:
       return IntelBroadwell;
+    case 0x406d0:
+      return IntelSilvermont;
     case 0x406e0:
     case 0x506e0:
       return IntelSkylake;
@@ -149,6 +153,7 @@ static void init_perf_event_attr(struct perf_event_attr* attr,
   // only.
   attr->exclude_kernel = 1;
   attr->exclude_guest = 1;
+  attr->precise_ip = 2;
 }
```
which is the `BR_INST_RETIRED.JCC`, but there still seems to be an overcount:
```
[FATAL /root/rr/src/ReplaySession.cc:763:check_ticks_consistency() errno: SUCCESS]
 (task 14414 (rec:14394) at time 210)
 -> Assertion `ticks_now == trace_ticks' failed to hold. ticks mismatch for 'SYSCALL: mprotect'; expected 48572, got 48573
```
I don't understand where this non-determinism comes from, but it always seems to be off-by-one. If somebody were to understand the cause of that, it may be correctable.

**Comment by Keno** [comment]
One more datapoint: Out of curiosity, I ran the code associated with the linked paper, which does seem to indicate that `0x517ec4` is deterministic (and in fact reproduces the value obtained by `0x5101c4` on ivybridge). So perhaps all hope is not lost.

**Comment by rocallahan** [comment]
You might want to check Silvermont errata since Intel sometimes documents miscounted events in errata.

**Comment by Keno** [comment]
Yes, the errata were the first thing I checked, but they do not mention this counter (they do mentioned the retired instruction counter).

**Comment by Keno** [comment]
I went back to this with a little better understanding of rr. I tried single stepping during replay and I can occasionally observe the tick counter increase by one even if it's not a branch instruction, or increase by 2 if it is a branch instruction.

**Comment by Keno** [comment]
Made some progress. By using cpuset to restrict other processes from running on the same core as the rr tracee, the replay got a lot further
```
root@scw-45a86a:~/julia/base# ~/rr-build/bin/rr replay -a
[FATAL /root/rr/src/ReplaySession.cc:751:check_ticks_consistency() errno: SUCCESS]
 (task 29837 (rec:29809) at time 244)
 -> Assertion `ticks_now == trace_ticks' failed to hold. ticks mismatch for 'SYSCALL: mprotect'; expected 937882, got 937881
Launch gdb with
  gdb /root/.local/share/rr/latest-trace/mmap_hardlink_1_llc
and attach to the rr debug server with:
  target remote :29837
```
and more importantly, the divergence is now reproducible (always at the same tick/ip/everything).
The code I used for cpuset was:
```
mkdir /dev/cpuset
mount -t cpuset cpuset /dev/cpuset
cd /dev/cpuset
mkdir default
mkdir rr
cd default
echo 0-6 > cpuset.cpus
while read i; do /bin/echo $i; done < ../tasks > tasks
cd ..
cd rr
echo 7 > cpuset.cpus
```

**Comment by khuey** [comment]
@Keno do you remember offhand if you were using a virtual or bare metal server for this?

**Comment by Keno** [comment]
Yes, this was bare metal. I am quite confident that this is a CPU bug since newer Silvermont based CPUs (e.g. KNL) do work with the above patch (I didn't get around to do much testing but most tests passed and simple rexord/replay was fine as well).

**Comment by rocallahan** [comment]
OK, that's useful to know!

**Comment by rocallahan** [comment]
Should we land your patch, then?

**Comment by Keno** [comment]
Yes, though we need to fill in the KNL cpuid. I should have that written down somewhere. Will find it when I'm back at an actual computer.

**Comment by Keno** [comment]
CPUID for the chip we tested on was `0x50670`.
